generic-extlinux-compatible/generic-extlinux-compatible.nix

```diff
@@ -0,0 +1,132 @@
+          under `/boot/extlinux.conf`.  For instance,
+          U-Boot's generic distro boot support uses this file format.
+
+          See [U-boot's documentation](https://u-boot.readthedocs.io/en/latest/develop/distro.html)
+          for more information.
+        '';
+      };
+
+      useGenerationDeviceTree = mkOption {
+        default = true;
+        type = types.bool;
+        description = ''
+          Whether to generate Device Tree-related directives in the
+          extlinux configuration.
+
+          When enabled, the bootloader will attempt to load the device
+          tree binaries from the generation's kernel.
+
+          Note that this affects all generations, regardless of the
+          setting value used in their configurations.
+        '';
+      };
+
+      configurationLimit = mkOption {
+        default = 20;
+        example = 10;
+        type = types.int;
+        description = ''
+          Maximum number of configurations in the boot menu.
+        '';
+      };
+
+      mirroredBoots = mkOption {
+        default = [ { path = "/boot"; } ];
+        example = [
+          { path = "/boot1"; }
+          { path = "/boot2"; }
+        ];
+        description = ''
+          Mirror the boot configuration to multiple paths.
+        '';
+
+        type = with types; listOf (submodule {
+          options = {
+            path = mkOption {
+              example = "/boot1";
+              type = types.str;
+              description = ''
+                The path to the boot directory where the extlinux-compatible
+                configuration files will be written.
+              '';
+            };
+          };
+        });
+      };
+
+      populateCmd = mkOption {
+        type = types.str;
+        readOnly = true;
+        description = ''
+          Contains the builder command used to populate an image,
+          honoring all options except the `-c <path-to-default-configuration>`
+          argument.
+          Useful to have for sdImage.populateRootCommands
+        '';
+      };
+
+      extraCommandsAfter = mkOption {
+        type = types.listOf types.str;
+        description = ''
+          Optional commands to run after installing the bootloader.
+          Useful for putting Raspberry Pi firmwares on the boot partition.
+        '';
+      };
+
+    };
+  };
+
+  config = let
+    builderArgs = "-g ${toString cfg.configurationLimit} -t ${timeoutStr}"
+      + lib.optionalString (dtCfg.name != null) " -n ${dtCfg.name}"
+      + lib.optionalString (!cfg.useGenerationDeviceTree) " -r";
+    installBootLoader = pkgs.writeScript "install-extlinux-conf.sh" (''
+      #!${pkgs.runtimeShell}
+      set -e
+    '' + flip concatMapStrings cfg.mirroredBoots (args: ''
+      ${builder} ${builderArgs} -d '${args.path}' -c "$@"
+    '') + ''
+      ${lib.concatLines cfg.extraCommandsAfter}
+    '');
+  in
+    mkIf cfg.enable {
+      system.build.installBootLoader = installBootLoader;
```

Comment:
@tstat you said:

> >  As part of this change, it was necessary to include a copied variant of generic-extlinux-compatible/*, which permits additional scripts to be added to system.build.installBootloader.
>
> Instead of copying and adding an option, I wonder if we can supply our own system.build.installBootloader that installs firmware files and calls generic-extlinux-compatible's installBootloader if uboot is enabled.

I'm not entirely sure I understand what you mean. I think this is what's happening on this line. The value of `config.system.build.installBootLoader` is different here than it would be in `generic-extlinux-compatible` as it now contains `cfg.extraCommandsAfter`. Can you please expand?